asset/asset.go

```diff
@@ -462,6 +464,66 @@ const (
 	ScriptV0 ScriptVersion = 0
 )
 
+// TapscriptTreePreimage represents the two supported ways to define a tapscript
+// tree to be used as a sibling for a Taproot Asset commitment, an asset group
+// key, or an asset script key.
+type TapscriptTreePreimage = fn.Either[TapBranchPreimage, []txscript.TapLeaf]
+
+func NewTreePreimageFromBranchPreimage(
+	b TapBranchPreimage) TapscriptTreePreimage {
+
+	return fn.NewLeft[TapBranchPreimage, []txscript.TapLeaf](b)
+}
+
+func NewTreePreimageFromLeaves(l []txscript.TapLeaf) (*TapscriptTreePreimage,
+	error) {
+
+	if len(l) == 0 {
+		return nil, fmt.Errorf("no leaves given")
+	}
+	// TODO(jhb): assert leaf versions?
+
+	preimage := fn.NewRight[TapBranchPreimage, []txscript.TapLeaf](l)
+
+	return &preimage, nil
+}
+
+// tapBranchPreimageLen is the length of a TapBranch preimage, excluding
+// the TapBranchTag.
+const TapBranchPreimageLen = 64
+
+type TapBranchPreimage struct {
+	left  [chainhash.HashSize]byte
+	right [chainhash.HashSize]byte
+}
+
+func NewTapBranchPreimage(branch txscript.TapBranch) TapBranchPreimage {
+	return TapBranchPreimage{
+		left:  branch.Left().TapHash(),
+		right: branch.Right().TapHash(),
+	}
+}
+
+func (b TapBranchPreimage) TapBranchHash() chainhash.Hash {
+	l, r := b.left[:], b.right[:]
```

Comment:
I think can lift the `tapBranchHash` function from elsewhere.

Reply:
I thought so too, but it was already reimplemented in the `commitment` package; the original in `txscript` is private.